teps/0138-decouple-api-and-feature-versioning.md

```diff
@@ -72,8 +91,331 @@ Per [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-
+
+| enable-api-fields |	legacy-enable-beta-features-by-default | enabled in `v1beta1`? | enabled in `v1`? |
+| ------ | ------ | --- | --- |
+| beta   |	true  |	yes |	yes |
+| beta   |	false |	yes |	yes |
+| stable |	true  |	no  |	no  |
+| stable |	false |	no  |	no  |
+
+Once all existing beta features become stable, `legacy-enable-beta-features-by-default` can be removed and we will deprecate and then remove `legacy-enable-beta-features-by-default` and to use `stable` `enable-api-fields`. We would default `true` for the new flag and after 9 months, we default `enable-api-fields` to `stable` to preserve the existing behavior.
+
+#### Pros
+- This will provide a smoother transition for switching default `enable-api-fields` value back to `stable`.
+
+#### Cons
+- This is a backwards incompatible change.
+- It is not clear what to do with alpha features for the feature promotion process when there are two `enable-api-fields` related flags, which might lead to confusions.
+- This is adding complexity to both the implementations and the testing matrix for the newly introduced flag.
+
+
+### Make validation changes only for new beta features
+This alternative proposes keeping beta features on by default. It will keep the current implementations and validations for beta features in `v1` and `v1beta1` apiVersion and only make new features have the synchronized validations across different apiVersions. This also proposes to keep the default value of `enable-api-fields` as `beta`,
+
+#### Pros
+- This is backwards compatible.
+- Changes are minimal for the current codebase with probably only documentation required.
+#### Cons
+- We would still need to use `beta` as the default stability of features that are turned on. Some beta features are still coupled in feature and API versioning. This will not allow cluster operators to opt-in only stable features in `v1beta1`.
+
+### Give 9-month warning before breaking changes and default to stable
+This alternative proposes the validation change for stable features in `v1beta1` to be done with giving a notice for the breaking change for 9 months.
+
+#### Cons:
+- Since `v1beta1` only has 12 months of support period left, making a breaking change to it might have more impacts on `v1beta1` users than its worth after 9 months.
+
+## Implementation Plan
+
+For each new feature driven by api fields, a new feature flag using FeatureFlag struct will be added to the existent set of feature flags. This struct will include the stability of feature and whether it is turned on or locked to default. Currently it will only work for new features without `enable-api-fields`, see future work for more details of the option to introduce `enable-api-fields=none` to use per feature flags for new features.
+
+
+```go
+type FeatureFlag struct {
+    // Name of the feature flag
+    Name string
+    
+    // Stability level of the feature, one of "stable", "beta" or "alpha"
+    Stability string
```

Comment:
This would basically only be for documentation purposes since the decision is only based on the `Enabled` part, so I'm not sure we need to have it part of the struct?

Reply:
This struct attempted to follow what [kubernetes has for their features](https://github.com/kubernetes/kubernetes/blob/56cc5e77a10ba156694309d9b6159d4cd42598e1/staging/src/k8s.io/component-base/featuregate/feature_gate.go#L73). And for the purposes of this field, I think there are two main benefits to have them:
- Previously we haven't had any real "source of truth" to tell what stability of a feature is apart from [additional-configs.md](https://github.com/tektoncd/pipeline/blob/release-v0.51.x/docs/additional-configs.md#alpha-features) which relies on manual edition and they can be moved to elsewhere sometimes (i recall once we had beta feature list in `install.md` , take [v0.41.x](https://github.com/tektoncd/pipeline/blob/release-v0.41.x/docs/install.md) for example). So with this field added, this should be regarded as the only source of truth for telling which stability level a feature is at currently.
- As pointed here, for easier documentation purposes. Maintainers/ Cluster operators would have a better sense of which stability level the specific feature is at.

Reply:
I think it would be interesting to see how that is used by k8s, but I guess it's ok to keep it even if we don't have a use for it yet.